{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/env python3\n",
    "\n",
    "import argparse\n",
    "import cairo\n",
    "import re\n",
    "\n",
    "# parser = argparse.ArgumentParser()\n",
    "# parser.add_argument(\"-f\", \"--filename\", required=True, help='Fasta file containing \\\n",
    "# single exons (capital) and flanking introns (lower-case)')\n",
    "# parser.add_argument(\"-m\", \"--motifs\", required=True, help='A file containing \\\n",
    "# query motifs (must be one sequence per line)')\n",
    "# args = parser.parse_args()\n",
    "        \n",
    "\n",
    "#Dictionary to translate between regex and sequence\n",
    "regex = {'A':'[Aa]','T':'[TtUu]','G':'[Gg]','C':'[Cc]','U':'[TtUu]',\n",
    "        'W':'[AaTtUu]','S':'[GgCc]','M':'[AaCc]','K':'[GgTtUu]','R':'[AaGg]','Y':'[CcTtUu]',\n",
    "        'B':'[CcGgTtUu]','D':'[AaGgTtUu]','H':'[AaCcTtUu]','V':'[AaCcGg]',\n",
    "        'N':'[GgCcAaTtUu]'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert input motifs to regex\n",
    "with open(\"motifs.txt\", 'r') as fh:\n",
    "    regex_motifs=[] # create empty list to output converted motifs\n",
    "    for line in fh:\n",
    "        regex_motif=''\n",
    "        motif=line.strip().upper()\n",
    "        for each in motif:\n",
    "            regex_motif+=regex[each]\n",
    "        regex_motifs.append(regex_motif)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dict to store gene lengths and exon position\n",
    "genes_index={}\n",
    "\n",
    "# Dict to store motif positions\n",
    "motifs_index={}\n",
    "\n",
    "def main_coords(sequence):\n",
    "    \"\"\"Input single fasta sequence containing capital EXON region.\n",
    "    Outputs list of [gene length, exon position, exon length].\"\"\"\n",
    "    gene_len = len(seq)\n",
    "    exon = re.search('[A-Z]',sequence)\n",
    "    exon_start = exon.start()\n",
    "    # Trim off upstream intron\n",
    "    sequence = sequence[exon_start:]\n",
    "    exon = re.search('[a-z]',sequence)\n",
    "    exon_len = exon.start()\n",
    "    out = [gene_len, exon_start, exon_len]\n",
    "    return out\n",
    "\n",
    "def motif_coords(sequence):\n",
    "    \"\"\"Input single fasta sequence containing capital EXON region.\n",
    "    Save positions of desired motifs.\"\"\"\n",
    "    out=[]\n",
    "    for motif in regex_motifs:\n",
    "        mot = re.finditer(motif,sequence)\n",
    "        for i in mot:\n",
    "            out.append(i.start())\n",
    "    return out\n",
    "        \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Main loop\n",
    "with open('sequence.txt','r') as fasta:\n",
    "    NR=0 #counts lines\n",
    "    for line in fasta:\n",
    "        NR+=1\n",
    "        # Save first header line\n",
    "        if NR == 1:\n",
    "            header=line.strip()\n",
    "            seq=''\n",
    "        # Save sequence\n",
    "        elif line.startswith('>') == False:\n",
    "            # Combines sequence lines if there are new line characters\n",
    "            seq=seq+line.strip()\n",
    "        # Save important info below\n",
    "        elif line.startswith('>'):\n",
    "            genes_index[header[1:]] = main_coords(seq)\n",
    "            motifs_index[header[1:]] = motif_coords(seq)\n",
    "            # Save next header and empty seq\n",
    "            header=line.strip()\n",
    "            seq=''\n",
    "            \n",
    "    # Final fasta record saved below\n",
    "    genes_index[header[1:]] = main_coords(seq)\n",
    "    motifs_index[header[1:]] = motif_coords(seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save dictionary key containing max gene length\n",
    "max_width = max(genes_index, key=genes_index.get)\n",
    "# Save max gene length\n",
    "max_width = genes_index[max_width][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set width based on max gene length and height based on gene count\n",
    "width,height = int(max_width*1.25), int(len(genes_index)*120)\n",
    "\n",
    "surface = cairo.SVGSurface(\"plot.svg\", width, height)\n",
    "context = cairo.Context(surface)\n",
    "\n",
    "# Useful coordinates\n",
    "x_start = width*0.1  # x position to start drawing each gene\n",
    "y_div = height/(len(genes_index)+1) # coefficient for y position of each gene\n",
    "\n",
    "# First draw genes as lines with solid rectangles for exons\n",
    "gene_cnt=1\n",
    "for val in genes_index.values():\n",
    "    gene_len=val[0] # length of gene\n",
    "    ex_start=val[1] # starting pos of exon\n",
    "    ex_end=ex_start+val[2] # ending pos of exon\n",
    "    \n",
    "    #draw gene line\n",
    "    context.set_line_width(2)\n",
    "    context.move_to(x_start, y_div*gene_cnt)\n",
    "    context.line_to(x_start+gene_len, y_div*gene_cnt)\n",
    "    context.stroke()\n",
    "    \n",
    "    #draw exon box\n",
    "    context.set_line_width(10)\n",
    "    context.move_to(x_start+ex_start, y_div*gene_cnt)\n",
    "    context.line_to(x_start+ex_end, y_div*gene_cnt)\n",
    "    context.stroke()\n",
    "    # increment index for next gene\n",
    "    gene_cnt+=1\n",
    "\n",
    "\n",
    "# Add lines indicating locations of motifs along genes\n",
    "gene_cnt=1\n",
    "for m_list in motifs_index.values():\n",
    "    for pos in m_list:\n",
    "        context.set_line_width(18)\n",
    "        context.set_source_rgb(250,0,0)\n",
    "        context.move_to(x_start+pos, y_div*gene_cnt)\n",
    "        context.line_to(x_start+pos+2, y_div*gene_cnt)\n",
    "        context.stroke()\n",
    "    gene_cnt+=1\n",
    "\n",
    "gene_cnt=1\n",
    "for name in genes_index:\n",
    "    #text label above gene\n",
    "    context.move_to(10, (y_div*gene_cnt)-20)\n",
    "    context.set_source_rgb(0,0,0)\n",
    "    context.select_font_face(\"Open Sans\", cairo.FONT_SLANT_NORMAL, cairo.FONT_WEIGHT_BOLD)\n",
    "    context.set_font_size(16)\n",
    "    context.show_text(name)\n",
    "    gene_cnt+=1\n",
    "\n",
    "# Add helpful info at the bottom\n",
    "context.move_to(10,height-15)\n",
    "context.set_font_size(12)\n",
    "context.show_text(\"*Black boxes represent exons. Red lines \\\n",
    "indicate locations of input motifs.\")\n",
    "\n",
    "\n",
    "surface.finish()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
